### PR TITLE
Yulproto funcdef stmt

### LINKED ISSUE #7276: [yul] proto fuzzer: Make function definition a statement
Move from top-level function definitions to those that can be declared anywhere as long as yul rules are respected.

### PR DESCRIPTION
**Pull request by bshastry**
fixes #7276

### CODE REVIEW

**Comment by chriseth** [comment]
please rebase.

**Comment by bshastry** [comment]
Rebased PR on top of target branch. Once the review passes, please merge this onto develop after #7309

File: test/tools/ossfuzz/protoToYul.h [hunk 1]
```diff
@@ -126,33 +113,56 @@ class ProtoConverter
...
+	/// - If we are inside a function that has already declared a variable
+	/// - If there is at least one variable declaration that is
+	/// in scope
+	bool varDeclAvailable();
+
+	/// Return true if a function call cannot be made, false otherwise.
+	/// @param _type is an enum denoting the type of function call. It
+	/// can be one of NONE, SINGLE, MULTIDECL, MULTIASSIGN.
+	/// 			NONE -> Function call does not return a value
+	///				SINGLE -> Function call returns a single value
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.h]: Is the difference in indentation on purpose?

File: test/tools/ossfuzz/protoToYul.cpp [hunk 2]
```diff
@@ -531,83 +528,134 @@ void ProtoConverter::visitFunctionInputParams(T const& _x, unsigned _numInputPar
...
+				m_output << ")\n";
+			}
+		}
 		break;
-	case FunctionCall::kCallMultidecl:
-		// Hack: Disallow (multi) variable declarations until scope extension is implemented for "for-init"
-		if (!m_inForInitScope)
-			visit(_x.call_multidecl());
+	case FunctionCall::SINGLE:
+		if (m_functionSigMap.size() > 0)
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp]: Can't this be combined with the above case?

File: test/tools/ossfuzz/protoToYul.cpp [hunk 3]
```diff
@@ -531,83 +528,134 @@ void ProtoConverter::visitFunctionInputParams(T const& _x, unsigned _numInputPar
...
+					m_output << ")\n";
+					// Add newly minted vars in the multidecl statement to current scope
+					addVarsToScope(varsVec);
+				}
+			}
+		}
 		break;
-	case FunctionCall::FUNCTIONCALL_ONEOF_NOT_SET:
+	case FunctionCall::MULTIASSIGN:
+		if (m_functionSigMap.size() > 0) {
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp]:
```suggestion
		if (m_functionSigMap.size() > 0)
		{
```

File: test/tools/ossfuzz/protoToYul.cpp [hunk 4]
```diff
@@ -531,83 +528,134 @@ void ProtoConverter::visitFunctionInputParams(T const& _x, unsigned _numInputPar
...
+				// Ensure that the chosen function returns at most 4 values
+				yulAssert(
+					numOutParams <= 4,
+					"Proto fuzzer: Function call with too many output params encountered."
+				);
+
+				// Convert LHS of multi assignment
+				// We reverse the order of out param visits since the order does not matter.
+				// This helps reduce the size of this switch statement.
+				switch (numOutParams) {
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp]:
```suggestion
				switch (numOutParams)
				{
```

File: test/tools/ossfuzz/protoToYul.cpp [hunk 5]
```diff
@@ -870,43 +921,90 @@ void ProtoConverter::visit(Statement const& _x)
...
 {
-	m_scopes.push({});
-	addToScope(_funcParams);
+	m_scopeVars.push({});
+	m_scopeFuncs.push({});
+	if (!_funcParams.empty())
+		addVarsToScope(_funcParams);
+}
+
+void ProtoConverter::updateFunctionMaps(string var)
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp]:
```suggestion
void ProtoConverter::updateFunctionMaps(string const& var)
```

File: test/tools/ossfuzz/protoToYul.cpp [hunk 6]
```diff
@@ -978,68 +1089,58 @@ void ProtoConverter::createFunctionDefAndCall(T const& _x, unsigned _numInParams
...
-}
+	m_inForBodyScope = wasInForBody;
+	m_inFunctionDef = wasInFunctionDef;
 
-void ProtoConverter::visit(FunctionDefinitionSingleReturnVal const& _x)
-{
-	unsigned numInParams = _x.num_input_params() % s_modInputParams;
-	unsigned numOutParams = 1;
-	createFunctionDefAndCall(_x, numInParams, numOutParams, NumFunctionReturns::Single);
+	// Manually create a multi assignment using global variables
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp]: Is this still needed?

Comment (bshastry) [reply test/tools/ossfuzz/protoToYul.cpp]: sorry, I meant to refactor the commented out portion. Will push an update.

**Comment by bshastry** [comment]
Updated PR.

**Comment by chriseth** [comment]
I think this PR is rather important to get in quickly, but I currently do not have the resources to look at it in depth. I'm fine to merge this if it generates valid and meaningful code.

**Comment by bshastry** [comment]
@chriseth I have been fuzzing this PR over the last couple of days locally and didn't find anything unexpected. For reference, here are some programs generated by it.

https://gist.github.com/bshastry/dde67b9239cef4ab71f953184a731705

I am currently investigating why test coverage slightly drops with this PR but I feel in the long run, this might not be a problem. Will update shortly with a more conclusive opinion.

**Comment by chriseth** [comment]
Ok, great!

Do you know why most of the programs are rather "uniform" - example: https://gist.github.com/bshastry/dde67b9239cef4ab71f953184a731705#file-100_yul_programs_generated_by_pr7318-L2196

**Review by chriseth — APPROVED** [review]

**Pull request merged**